src/object/cli_obj.c

```diff
@@ -1462,49 +1423,521 @@ shard_task_sched(tse_task_t *task, void *arg)
+		}
+		/* decref and delete from head at shard_task_remove */
+		tse_task_addref(shard_task);
+		tse_task_list_add(shard_task, task_list);
+	}
+
+task_sched:
+	obj_shard_task_sched(obj_auxi);
+	return 0;
+
+out_task:
+	if (d_list_empty(task_list))
+		tse_task_complete(obj_task, rc);
+	else
+		tse_task_list_traverse(task_list, shard_task_abort, &rc);
+	return rc;
+}
+
+static int
+shard_result_process(tse_task_t *task, void *arg)
+{
+	struct obj_auxi_args	*obj_auxi = (struct obj_auxi_args *)arg;
+	struct shard_auxi_args	*shard_auxi;
+	int			 ret = task->dt_result;
+
+	/*
+	 * Check shard IO task's completion status:
+	 * 1) if succeed just stores the highest replied pm version.
+	 * 2) if any shard got retryable error, mark the IO as retryable.
+	 * 3) for the un-retryable failure, store it in obj_auxi->result.
+	 */
+	shard_auxi = tse_task_buf_embedded(task, sizeof(*shard_auxi));
+	if (ret == 0) {
+		if (obj_auxi->map_ver_reply < shard_auxi->map_ver)
+			obj_auxi->map_ver_reply = shard_auxi->map_ver;
+	} else if (obj_retry_error(ret)) {
+		D_DEBUG(DB_IO, "shard %d ret %d.\n", shard_auxi->shard, ret);
+		obj_auxi->io_retry = 1;
+		if (ret == -DER_INPROGRESS)
+			obj_auxi->retry_with_leader = 1;
+	} else {
+		/* for un-retryable failure, set the err to whole obj IO */
+		D_DEBUG(DB_IO, "shard %d ret %d.\n", shard_auxi->shard, ret);
+		obj_auxi->result = ret;
+	}
+
+	return 0;
+}
+
+static int
+shard_task_remove(tse_task_t *task, void *arg)
+{
+	tse_task_list_del(task);
+	tse_task_decref(task);
+	return 0;
+}
+
+static void
+shard_task_list_init(struct obj_auxi_args *auxi)
+{
+	if (auxi->io_retry == 0) {
+		D_INIT_LIST_HEAD(&auxi->shard_task_head);
+		return;
+	}
+}
+
+static void
+obj_list_dkey_cb(tse_task_t *task, struct obj_list_arg *arg, unsigned int opc)
+{
+	struct dc_object       *obj = arg->obj;
+	daos_anchor_t	       *anchor = arg->dkey_anchor;
+	uint32_t		shard = dc_obj_anchor2shard(anchor);
+	int			grp_size;
+
+	if (task->dt_result != 0)
+		return;
+
+	grp_size = obj_get_grp_size(obj);
+	D_ASSERT(grp_size > 0);
+
+	if (!daos_anchor_is_eof(anchor)) {
+		D_DEBUG(DB_IO, "More keys in shard %d\n", shard);
+	} else if ((shard < obj->cob_shards_nr - grp_size)) {
+		shard += grp_size;
+		D_DEBUG(DB_IO, "next shard %d grp %d nr %u\n",
+			shard, grp_size, obj->cob_shards_nr);
+
+		daos_anchor_set_zero(anchor);
+		dc_obj_shard2anchor(anchor, shard);
+	} else {
+		D_DEBUG(DB_IO, "Enumerated All shards\n");
+	}
+}
+
+static int
+obj_comp_cb(tse_task_t *task, void *data)
+{
+	struct obj_list_arg	*arg;
+	struct dc_object	*obj;
+	struct obj_auxi_args	*obj_auxi;
+	struct shard_auxi_args	*shard_auxi;
+	d_list_t		*head = NULL;
+	bool			 pm_stale = false;
+
+	obj_auxi = tse_task_stack_pop(task, sizeof(*obj_auxi));
+	obj_auxi->retry_with_leader = 0;
+	obj_auxi->io_retry = 0;
+	switch (obj_auxi->opc) {
+	case DAOS_OBJ_DKEY_RPC_ENUMERATE:
+		arg = data;
+		obj = arg->obj;
+		obj_list_dkey_cb(task, arg, obj_auxi->opc);
+		break;
+	case DAOS_OBJ_RPC_ENUMERATE:
```

Comment:
Should it be for DAOS_OBJ_DKEY_RPC_ENUMERATE?

Reply:
DAOS_OBJ_DKEY_RPC_ENUMERATE is at above L1712.
I did not change this part actually just moved it ahead.

Reply:
I am some confused, why DAOS_OBJ_RPC_ENUMERATE check "if (daos_anchor_is_eof(arg->dkey_anchor))" at line 1720? It may be not your patch issue. Why not is the DAOS_OBJ_DKEY_RPC_ENUMERATE?

Reply:
it is just to add a log, I think it is fine to also add such as check and log to DAOS_OBJ_DKEY_RPC_ENUMERATE.
Why check "if (daos_anchor_is_eof(arg->dkey_anchor))" for DAOS_OBJ_RPC_ENUMERATE is that for rebuild it enumerates dkey/akey/recxs recursively (so it pass all anchor/dkey_anchor/akey_anchor 3 anchor). When that check is true it means the recursively enumerate is done because the first-level enumerate is for specific object's dkey. just my understanding, @wangdi1 is more familiar with this part.